libbeat/common/transport/tlscommon/tls_config.go

```diff
@@ -86,68 +87,177 @@ func (c *TLSConfig) ToConfig() *tls.Config {
+		ClientAuth:         c.ClientAuth,
+		Time:               c.time,
+		VerifyConnection:   verifyConnectionFn,
 	}
 }
 
 // BuildModuleConfig takes the TLSConfig and transform it into a `tls.Config`.
-func (c *TLSConfig) BuildModuleConfig(host string) *tls.Config {
+func (c *TLSConfig) BuildModuleClientConfig(host string) *tls.Config {
+	if c == nil {
+		// use default TLS settings, if config is empty.
+		return &tls.Config{
+			ServerName:         host,
+			InsecureSkipVerify: true,
+			VerifyConnection: makeVerifyConnection(&TLSConfig{
+				Verification: VerifyFull,
+			}),
+		}
+	}
+
+	config := c.ToConfig()
+	config.ServerName = host
+	config.VerifyConnection = makeVerifyConnection(c)
+	return config
+}
+
+// BuildModuleConfig takes the TLSConfig and transform it into a `tls.Config`.
+func (c *TLSConfig) BuildModuleServerConfig(host string) *tls.Config {
 	if c == nil {
 		// use default TLS settings, if config is empty.
-		return &tls.Config{ServerName: host}
+		return &tls.Config{
+			ServerName:         host,
+			InsecureSkipVerify: true,
+			VerifyConnection: makeVerifyServerConnection(&TLSConfig{
+				Verification: VerifyFull,
+			}),
+		}
 	}
 
 	config := c.ToConfig()
 	config.ServerName = host
+	config.VerifyConnection = makeVerifyServerConnection(c)
 	return config
 }
 
-// makeVerifyPeerCertificate creates the verification combination of checking certificate pins and skipping host name validation depending on the config
-func makeVerifyPeerCertificate(cfg *TLSConfig) verifyPeerCertFunc {
+func makeVerifyConnection(cfg *TLSConfig) func(tls.ConnectionState) error {
 	pin := len(cfg.CASha256) > 0
-	skipHostName := cfg.Verification == VerifyCertificate
 
-	if pin && !skipHostName {
-		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-			return verifyCAPin(cfg.CASha256, verifiedChains)
+	switch cfg.Verification {
+	case VerifyFull:
+		return func(cs tls.ConnectionState) error {
+			// On the client side, PeerCertificates can't be empty.
+			if len(cs.PeerCertificates) == 0 {
+				return fmt.Errorf("no peer certificates")
+			}
+
+			dnsnames := cs.PeerCertificates[0].DNSNames
+			var serverName string
+			if len(dnsnames) == 0 || len(dnsnames) == 1 && dnsnames[0] == "" {
+				serverName = cs.PeerCertificates[0].Subject.CommonName
+			} else {
+				serverName = dnsnames[0]
+			}
+			if len(serverName) > 0 && len(cs.ServerName) > 0 && serverName != cs.ServerName {
+				return x509.HostnameError{Certificate: cs.PeerCertificates[0], Host: cs.ServerName}
+			}
+			opts := x509.VerifyOptions{
+				Roots:         cfg.RootCAs,
+				Intermediates: x509.NewCertPool(),
+			}
+			for _, cert := range cs.PeerCertificates[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			verifiedChains, err := cs.PeerCertificates[0].Verify(opts)
+			if err != nil {
+				return err
+			}
+
+			if pin {
+				return verifyCAPin(cfg.CASha256, verifiedChains)
+			}
+			return nil
 		}
-	}
+	case VerifyCertificate:
```

Comment:
This looks almost equivalent to the `Full` mode (besides the hostname check). How about moving the host verification into a separate function and us combining the two functions into one?